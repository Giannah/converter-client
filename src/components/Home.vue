<template>
  <v-container>
    <v-row justify="center">
      <v-dialog v-model="isLoading" persistent width="300" dark>
        <div class="text-center">
          <v-card color="indigo" dark>
            <v-card-title>Converting... </v-card-title>
            <v-card-text>
              <v-progress-circular
                indeterminate
                size="70"
                :width="7"
                color="white"
              ></v-progress-circular>
            </v-card-text>
          </v-card>
        </div>
      </v-dialog>
    </v-row>

    <v-row class="text-center">
      <v-col cols="12">
        <v-img :src="require('../assets/romancharacter.svg')" class="my-1" contain height="300" />
      </v-col>

      <v-col class="mb-4">
        <h1 class="display-2 font-weight-bold mb-3">Welcome to the AToR Converter !</h1>

        <p class="subheading font-weight-regular">
          Simply type in an arabic numeral from 1 to 100 and get it converted to it's roman numeral
          equivalent.
        </p>
      </v-col>

      <v-col v-if="convertedNumber.length > 0" class="mb-5" cols="12">
        <converted-alert
          :convertedNumber="convertedNumber"
          @isClicked="handleClick"
        ></converted-alert>
      </v-col>

      <v-col v-if="errorMessage.length > 0" class="mb-5" cols="12">
        <error-alert :errorMessage="errorMessage" @isClicked="handleErrorMessage"></error-alert>
      </v-col>

      <v-col class="mb-5" cols="12" v-show="!convertedNumber && !errorMessage">
        <h2 class="headline font-weight-bold mb-3">Enter your number here ...</h2>

        <v-row justify="center">
          <v-col cols="12" sm="6" md="3">
            <v-form
              class="d-flex flex-row flex-nowrap align-baseline justify-space-around"
              ref="form"
            >
              <v-text-field
                data-test="input"
                v-model="value"
                class="mx-6 py-2"
                type="number"
                label="Arabic numeral"
                min="1"
                max="100"
                step="1"
                :rules="valueRules"
                @keydown.enter="submitForm"
                required
                outlined
              ></v-text-field>
              <v-btn color="indigo" fab dark class="py-2" @click="submitForm">Go !</v-btn>
            </v-form>
          </v-col>
        </v-row>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
import getRomanConversionService from '@/services/ConverterService.js'
import ConvertedAlert from '@/components/ConvertedAlert.vue'
import ErrorAlert from '@/components/ErrorAlert.vue'
import axios from 'axios'
export default {
  name: 'Home',
  components: {
    ConvertedAlert,
    ErrorAlert,
  },
  data: () => ({
    value: '',
    isLoading: false,
    convertedNumber: '',
    errorMessage: '',
    valueRules: [
      (v) => !!v || 'Value is required',
      (v) => (v && /^[1-9][0-9]?$|^100$/gm.test(v)) || 'Value must be between 1 and 100',
    ],
  }),
  methods: {
    submitForm() {
      const apiClient = axios.create({
        baseURL: 'http://localhost:5000/rest/converter',
        withCredentials: false,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
      })
      const ConverterService = getRomanConversionService(apiClient)
      if (this.$refs.form.validate()) {
        this.isLoading = true
        ConverterService.getConvertedNumber({ input: this.value })
          .then((response) => {
            this.convertedNumber = response.data
            this.isLoading = false
          })
          .catch((error) => {
            this.errorMessage = `${error}`
            this.isLoading = false
          })
        this.$refs.form.reset()
      }
    },

    handleClick() {
      this.convertedNumber = ''
    },

    handleErrorMessage() {
      this.errorMessage = ''
    },
  },
}
</script>
<style lang="scss" scoped>
.loading-dialog {
  background-color: #303030;
}
</style>
