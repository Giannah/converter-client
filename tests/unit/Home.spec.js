import Vue from 'vue'
import Vuetify from 'vuetify'

import getRomanConversionService from '@/services/ConverterService.js'

import Home from '@/components/Home.vue'

import { createLocalVue, mount } from '@vue/test-utils'

Vue.use(Vuetify)
const localVue = createLocalVue()

describe('Home', () => {
  let vuetify
  let apiClient
  let ConverterService
  let wrapper

  const mockGetConvertedNumber = jest.fn()

  apiClient = {
    post: Promise.resolve({ data: 'I' }),
  }

  ConverterService = getRomanConversionService(apiClient)

  jest.mock('@/services/ConverterService.js', () => {
    return jest.fn().mockImplementation(() => {
      mockGetConvertedNumber
    })
  })

  beforeEach(() => {
    vuetify = new Vuetify()
    wrapper = mount(Home, {
      localVue,
      vuetify,
      apiClient,
      ConverterService,
    })
  })

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Input', async () => {
    const input = wrapper.find('[data-test="input"]')
    await input.setValue('1')

    expect(input.element.value).toBe('1')
  })

  it('triggers submit event', async () => {
    const input = wrapper.find('[data-test="input"]')
    await input.setValue('1')

    const button = wrapper.find('.v-btn')

    button.trigger('click')
  })
})
