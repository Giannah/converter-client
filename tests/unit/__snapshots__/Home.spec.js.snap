// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<div class="container">
  <div class="row justify-center">
    <div role="dialog" class="v-dialog__container">
      <!---->
    </div>
  </div>
  <div class="row text-center">
    <div class="col col-12">
      <div class="v-image v-responsive my-1 theme--light" style="height: 300px;">
        <div class="v-responsive__content"></div>
      </div>
    </div>
    <div class="mb-4 col">
      <h1 class="display-2 font-weight-bold mb-3">Welcome to the AToR Converter !</h1>
      <p class="subheading font-weight-regular">
        Simply type in an arabic numeral from 1 to 100 and get it converted to it's roman numeral
        equivalent.
      </p>
    </div>
    <!---->
    <div class="mb-5 col col-12">
      <h2 class="headline font-weight-bold mb-3">Enter your number here ...</h2>
      <div class="row justify-center">
        <div class="col-sm-6 col-md-3 col-12">
          <form novalidate="novalidate" class="v-form d-flex flex-row flex-nowrap align-baseline justify-space-around">
            <div class="v-input mx-6 py-2 theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <fieldset aria-hidden="true">
                    <legend style="width: 0px;"><span>â€‹</span></legend>
                  </fieldset>
                  <div class="v-text-field__slot"><label for="input-5" class="v-label theme--light" style="left: 0px; position: absolute;">Arabic numeral</label><input data-test="input" min="1" max="100" step="1" required="required" id="input-5" type="number"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div> <button type="button" class="py-2 v-btn v-btn--is-elevated v-btn--fab v-btn--has-bg v-btn--round theme--dark v-size--default indigo"><span class="v-btn__content">Go !</span></button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
